lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.upsertWithWhere(modelName, where, update, done);
+          });
+        }
+        function done(err, data, info) {
+          if (err) return cb(err);
+          var context = {
+            Model: Model,
+            data: data,
+            isNewInstance: info && info.isNewInstance,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('loaded', context, function(err) {
+            if (err) return cb(err);
+            var obj;
+            if (data && !(data instanceof Model)) {
+              inst._initProperties(data, { persisted: true });
+              obj = inst;
+            } else {
+              obj = data;
+            }
+            if (err) {
+              cb(err, obj);
+            } else {
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            }
+          });
+        }
+      });
+    } else {
+      var element = Object.keys(where)[0];
+      var filter = { 'where': where };
+      if (!Object.keys(data).length) {
+        process.nextTick(function() {
+          cb('data field cannot be empty');
+        });
+        return cb.promise;
+      }
 
+      if (!data.hasOwnProperty(element)) {
+        data[element] = where[element];
+      }
+      var pk = idName(Model);
+      self.find(filter, function(err, instances) {
+        var modelsLength = instances.length;
+        if (modelsLength === 0 && data.hasOwnProperty(pk)) {
+          self.create(data, options, cb);
+          return cb.promise;
+        } else if (modelsLength == 0 && data.hasOwnProperty(pk) == false) {
+          cb('Error message: No matching instance found.Found error while trying ' +
+                'to create a new instance.' +
+                'Primary key and value ' +
+                'is not provided in the payload');
+        } else if (modelsLength === 1) {
+          var pkName = idName(Model);
+          var idVal = instances[0][pkName];
+          var modelInst = instances[0];
+          modelInst.updateAttributes(data, function(err, info) {
+            if (err) {
+              cb(err);
+            } else {
+              cb(null, info);
+            }
+          });
+          return cb.promise;
+        } else {
+          process.nextTick(function() {
+            cb('There are multiple instances found.Upsert Operation will not be performed');
+          });
+        }
+      });
+      return cb.promise;
```

Comment:
In fact even this is duplicate; I believe `return cb.promise;` on line 825 is enough!